modules/core/src/main/java/org/apache/synapse/util/xpath/SynapseXPath.java

```diff
@@ -368,7 +372,15 @@ public String stringValueOf(MessageContext synCtx) {
                             ((Axis2MessageContext) synCtx).getEnvelope().getBody().getFirstElement() == null)) {
                 try {
                     axis2MC = ((Axis2MessageContext) synCtx).getAxis2MessageContext();
-                    inputStream = getMessageInputStreamPT(axis2MC);
+                    String contentType = (String) axis2MC.getProperty(SynapseConstants.AXIS2_PROPERTY_CONTENT_TYPE);
+                    if (!isStreamingXpathSupportedContentType(contentType) &&
+                            !Boolean.TRUE.equals(PassThroughConstants.MESSAGE_BUILDER_INVOKED)) {
+                        RelayUtils.buildMessage(axis2MC);
+                    } else {
+                        inputStream = getMessageInputStreamPT(axis2MC);
+                    }
+                } catch (XMLStreamException e) {
+                    handleException("Error occurred while building the message from the message context", e);
                 } catch (IOException e) {
                     log.error("Error occurred while obtaining input stream from the message context", e);
```

Comment:
Don't we need to handle IOExceptions?

Reply:
Not related to this fix, but we can improve rest while fixing

Reply:
If we handle the exception here, the flow will be stopped and faultHandler will be executed. Even IOException has occurred, flow should be continued because Xpath evaluation will be continued over existing SOAP envelope when the input stream is null.